src/audio/asrc/asrc.h

```diff
@@ -0,0 +1,284 @@
+	 */
+	IPC4_MOD_ASRC_PUSH_MODE = 0,
+	/* Use for capture/downlink
+	 */
+	IPC4_MOD_ASRC_PULL_MODE = 1,
+	/* Jitter buffer in ASRC implementation is optional and can be disabled
+	 * when using DP mode. For keeping backwards compatibility, by default
+	 * Jitter buffer is enabled
+	 */
+	IPC4_MOD_ASRC_DISABLE_JITTER_BUFFER = 4
+};
+
+struct ipc4_asrc_module_cfg {
+	struct ipc4_base_module_cfg base;
+	/* ASRC output sampling frequency. */
+	enum ipc4_sampling_frequency out_freq;
+
+	/* Mask of allowed ASRC features:
+	 * - BITS 0-1 (ASRC mode):
+	 *	a. 0 1 - Playback mode / PUSH mode
+	 *	b. 1 0 - Capture mode / PULL mode
+	 *
+	 * - BIT 4 (Disable Jitter Buffer):
+	 *  a. 0 - Jitter buffer enabled
+	 *  b. 1 - Jitter buffer disabled
+	 */
+	uint32_t asrc_mode;
+} __packed __aligned(4);
+#endif
+
+/* Simple count value to prevent first delta timestamp
+ * from being input to low-pass filter.
+ */
+#define TS_STABLE_DIFF_COUNT	2
+
+/* Low pass filter coefficient for measured drift factor,
+ * The low pass function is y(n) = c1 * x(n) + c2 * y(n -1)
+ * coefficient c2 needs to be 1 - c1.
+ */
+#define COEF_C1		Q_CONVERT_FLOAT(0.01, 30)
+#define COEF_C2		Q_CONVERT_FLOAT(0.99, 30)
+
+#ifdef CONFIG_IPC_MAJOR_4
+typedef struct ipc4_asrc_module_cfg ipc_asrc_cfg;
+#else
+typedef struct ipc_config_asrc ipc_asrc_cfg;
+#endif
+
+typedef void (*asrc_proc_func)(struct processing_module *mod,
+			       const struct audio_stream *source,
+			       struct audio_stream *sink,
+			       int *consumed,
+			       int *produced);
+/* asrc component private data */
+struct comp_data {
+	ipc_asrc_cfg ipc_config;
+	struct asrc_farrow *asrc_obj;	/* ASRC core data */
+	struct comp_dev *dai_dev;	/* Associated DAI component */
+	enum asrc_operation_mode mode;  /* Control for push or pull mode */
+	uint64_t ts;
+	uint32_t sink_rate;	/* Sample rate in Hz */
+	uint32_t source_rate;	/* Sample rate in Hz */
+	uint32_t sink_format;	/* For used PCM sample format */
+	uint32_t source_format;	/* For used PCM sample format */
+	uint32_t copy_count;	/* Count copy() operations  */
+	int32_t ts_prev;
+	int32_t sample_prev;
+	int32_t skew;		/* Rate factor in Q2.30 */
+	int32_t skew_min;
+	int32_t skew_max;
+	int ts_count;
+	int asrc_size;		/* ASRC object size */
+	int buf_size;		/* Samples buffer size */
+	int frames;		/* IO buffer length */
+	int source_frames;	/* Nominal # of frames to process at source */
+	int sink_frames;	/* Nominal # of frames to process at sink */
+	int source_frames_max;	/* Max # of frames to process at source */
+	int sink_frames_max;	/* Max # of frames to process at sink */
+	int data_shift;		/* Optional shift by 8 to process S24_4LE */
+	uint8_t *buf;		/* Samples buffer for input and output */
+	uint8_t *ibuf[PLATFORM_MAX_CHANNELS];	/* Input channels pointers */
+	uint8_t *obuf[PLATFORM_MAX_CHANNELS];	/* Output channels pointers */
+	bool track_drift;
+	asrc_proc_func asrc_func;		/* ASRC processing function */
+};
+
+#ifndef CONFIG_IPC_MAJOR_4
```

Comment:
do you think below is a good method?
combine all CONFIG_IPC_MAJOR_4 CONFIG_IPC_MAJOR_3
with only once in header file, and move all related inside.
#if CONFIG_IPC_MAJOR_4

#8034 
#if CONFIG_IPC_MAJOR_3
#endif

any comments?

Reply:
This is fine for now. We can incrementally update if needed.